test-data/unit/check-overloading.test

```diff
@@ -1723,6 +1743,105 @@ def foo2(**kwargs: int) -> str: ...
 def foo2(*args: int) -> int: ...  # E: Overloaded function signature 2 will never be matched: function 1's parameter type(s) are the same or broader
 [builtins fixtures/dict.pyi]
 
+[case testOverloadVarargInputAndVarargDefinition]
+from typing import overload, List
+
+class A: ...
+class B: ...
+class C: ...
+
+@overload
+def foo(x: int) -> A: ...
+@overload
+def foo(x: int, y: int) -> B: ...
+@overload
+def foo(x: int, y: int, z: int, *args: int) -> C: ...
```

Comment:
I would add a test with overload like this:
```python
@overload
def f(x: int) -> Tuple[int]: ...
@overload
def f(x: int, y: int) -> Tuple[int, int]: ...
@overload
def f(*xs: int) -> Tuple[int, ...]: ...
```
And check how it various calls with and without star.